src/main/java/com/ontotext/kafka/processor/SinkRecordsProcessor.java

```diff
@@ -0,0 +1,320 @@
+					retryOperator.execute(() -> handleRecord(record, connection), Stage.KAFKA_CONSUME, getClass());
+					if (retryOperator.failed()) {
+						LOG.warn("Failed to commit record. Will handle failure, and remove from the batch");
+						errorHandler.handleFailingRecord(record, retryOperator.error());
+						if (!retryOperator.withinToleranceLimits()) {
+							throw new ConnectException("Error tolerance exceeded.");
+						}
+					}
+					recordsBatch.poll();
+				}
+
+			}
+			connection.commit();
+			LOG.trace("Transaction commited, Batch size: {} , Records in current batch: {}", batchSize, recordsInCurrentBatch);
+			if (LOG.isTraceEnabled()) {
+				long finish = System.currentTimeMillis();
+				LOG.trace("Finished batch processing for {} ms", finish - start);
+			}
+			return null;
+		} catch (RepositoryException e) {
+			throw new RetriableException(e);
+		}
+	}
+
+	Void handleRecord(SinkRecord record, RepositoryConnection connection) {
+		long start = System.currentTimeMillis();
+		try {
+			LOG.trace("Executing {} operation......", transactionType.toString().toLowerCase());
+			recordHandler.handle(record, connection);
+			return null;
+		} catch (IOException e) {
+			throw new RetriableException(e.getMessage(), e);
+		} finally {
+			if (LOG.isTraceEnabled()) {
+				LOG.trace("Record info: {}", ValueUtil.recordInfo(record));
+				long finish = System.currentTimeMillis();
+				LOG.trace("Converted the record and added it to the RDF4J connection for {} ms", finish - start);
+			}
+		}
+
+	}
+
+	/**
+	 * Handles incoming {@link SinkRecord} based on specified @link {@link com.ontotext.kafka.GraphDBSinkConfig.TransactionType}.
+	 * Note: This is not NPE-safe, see @{@link ValueUtil#convertValueToString(Object)}
+	 *
+	 * @param transactionType Transaction type to use for record handling operation
+	 * @return The handler
+	 */
+	private RecordHandler getRecordHandler(GraphDBSinkConfig.TransactionType transactionType) {
+		switch (transactionType) {
+			case ADD:
+				return (record, connection) -> {
+					connection.add(ValueUtil.convertRDFData(record.value()), format);
+				};
+
+			case SMART_UPDATE:
+				return (record, connection) -> {
+					String query = getUpdateQuery(record.key());
+					connection.prepareUpdate(query).execute();
+					connection.add(ValueUtil.convertRDFData(record.value()), format);
+				};
+
+			case REPLACE_GRAPH:
+				return (record, connection) -> {
+					Resource context = ValueUtil.convertIRIKey(record.key());
+					connection.clear(context);
+					LOG.trace("Connection cleared context(IRI): {}", context.stringValue());
+					if (record.value() != null) {
+						connection.add(ValueUtil.convertRDFData(record.value()), format, context);
+					}
+				};
+			default:
+				throw new NotImplementedException(String.format("Handler for transaction type %s not implemented", transactionType));
+
+		}
+	}
+
+	private String getUpdateQuery(Object key) {
```

Comment:
If we move out the different operations- add / replace graph / smart update, this can go in the smart update logic

Reply:
I've moved it to the `ValueUtil.java` and renamed it `createUpdateRecordQuery`. What do you think about that approach? 
In this sense, the 3 ops become pretty straightforward (moved to a separate interface/factory)